modules/ismp/clients/polygon/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+
+		if let Some(milestone_update_ref) = &polygon_consensus_update.milestone_update {
+			// Check milestone_number == milestone.end_block and both match evm_header.number
+			let end_block =
+				milestone_update_ref.milestone.end_block.parse::<u64>().map_err(|_| {
+					ismp::error::Error::Custom("Invalid milestone.end_block".to_string())
+				})?;
+			let evm_block = milestone_update_ref.evm_header.number.low_u64();
+			if milestone_update_ref.milestone_number != end_block || end_block != evm_block {
+				return Err(ismp::error::Error::Custom(
+					"Milestone number, end_block, and EVM header number do not match".to_string(),
+				));
+			}
+
+			if milestone_update_ref.milestone_number <= consensus_state.last_finalized_block {
+				return Err(ismp::error::Error::Custom("Expired update".to_string()));
+			}
+
+			let commitment_proof = ics23::CommitmentProof::decode(
+				&mut &milestone_update_ref.ics23_state_proof.proof[..],
+			)
+			.map_err(|e| ismp::error::Error::Custom(e.to_string()))?;
+
+			let spec = ics23::tendermint_spec();
+
+			let verification_result = ics23::verify_membership::<HostFunctionsManager>(
+				&commitment_proof,
+				&spec,
+				&milestone_update_ref.untrusted_header_app_hash,
+				&milestone_update_ref.ics23_state_proof.key,
+				&milestone_update_ref.ics23_state_proof.value,
+			);
+
+			if !verification_result {
+				return Err(ismp::error::Error::Custom(
+					"ICS23 proof verification failed".to_string(),
+				));
+			}
+		}
+
+		let consensus_proof = polygon_consensus_update
+			.tendermint_proof
+			.to_consensus_proof()
+			.map_err(|e| ismp::error::Error::Custom(e.to_string()))?;
+
+		let trusted_state = CodecTrustedState::decode(&mut &consensus_state.tendermint_state[..])
+			.map_err(|e| ismp::error::Error::Custom(e.to_string()))?
+			.to_trusted_state()
+			.map_err(|e| ismp::error::Error::Custom(e.to_string()))?;
+
+		let time = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs();
+
+		let result = verify_header_update(trusted_state, consensus_proof, time);
+
+		match result {
+			Ok(updated_state) => {
+				let mut state_machine_map: BTreeMap<StateMachineId, Vec<StateCommitmentHeight>> =
+					BTreeMap::new();
+
+				if let Some(milestone_update) = &milestone_update {
+					let evm_header = &milestone_update.evm_header;
+
+					let state_commitment = StateCommitmentHeight {
+						commitment: StateCommitment {
+							timestamp: evm_header.timestamp,
+							overlay_root: None,
+							state_root: evm_header.state_root,
+						},
+						height: evm_header.number.low_u64(),
+					};
+
+					state_machine_map.insert(
+						StateMachineId {
+							state_id: StateMachine::Evm(POLYGON_CHAIN_ID),
+							consensus_state_id: _consensus_state_id,
+						},
+						vec![state_commitment],
+					);
+				}
+
+				let mut updated_consensus_state = consensus_state.clone();
+				updated_consensus_state.last_finalized_block = updated_state.verified_height;
```

Comment:
This value should be updated to the evm_header.number
```suggestion
				updated_consensus_state.last_finalized_block = updated_state.verified_height;
```